compiler/dcalc/invariants.ml

```diff
@@ -121,6 +122,103 @@ let invariant_match_inversion () : string * invariant_expr =
         else Fail
       | _ -> Ignore )
 
+(* The purpose of these functions is to determine whether the type `TDefault`
+   can only appear in certain positions, such as:
+
+   * On the left-hand side of an arrow with arity 1, as the type of a scope (for
+   scope calls).
+
+   * At the root of the type tree (outside a default).
+
+   * On the right-hand side of the arrow at the root of the type (occurs for
+   rentrant variables).
+
+   For instance, the following types do follow the invariant:
+
+   int; bool; int -> bool; <bool>; <int -> bool>; int -> <bool>; S_in {x: int ->
+   <bool>} -> S {y: bool}
+
+   While the following types does not follow the invariant:
+
+   <<int>>; <int -> <bool>>; <bool> -> int; S_in {x: int -> <bool>} -> S {y:
+   <bool>}
+
+   This is crucial to maintain the safety of the type system, as demonstrated in
+   the formal development. *)
+
+let rec check_typ_no_default ctx ty =
+  match Mark.remove ty with
+  | TLit _ -> true
+  | TTuple ts -> List.for_all (check_typ_no_default ctx) ts
+  | TStruct n ->
+    let s = StructName.Map.find n ctx.ctx_structs in
+    StructField.Map.for_all (fun _k ty -> check_typ_no_default ctx ty) s
+  | TEnum n ->
+    let s = EnumName.Map.find n ctx.ctx_enums in
+    EnumConstructor.Map.for_all (fun _k ty -> check_typ_no_default ctx ty) s
+  | TOption ty -> check_typ_no_default ctx ty
+  | TArrow (args, res) ->
+    List.for_all (check_typ_no_default ctx) args && check_typ_no_default ctx res
+  | TArray ty -> check_typ_no_default ctx ty
+  | TDefault _t -> false
+  | TAny ->
+    (* found TAny *)
+    Message.emit_warning
```

Comment:
`emit_warning "Internal error"` seems contradictory.
I believe this could be just a debug message ? After all it might be legitimate to want to check what can be checked in specific passes...